File: go/148/main.go
package main

import "fmt"

func main() {
	h1 := &ListNode{
		Val: 4,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 1,
				Next: &ListNode{
					Val:  3,
					Next: nil,
				},
			},
		},
	}

	h2 := &ListNode{
		Val: -1,
		Next: &ListNode{
			Val: 5,
			Next: &ListNode{
				Val: 3,
				Next: &ListNode{
					Val: 4,
					Next: &ListNode{
						Val:  0,
						Next: nil,
					},
				},
			},
		},
	}

	h3 := (*ListNode)(nil)

	// Expecting [1, 2, 3, 4]
	printLinkedList(sortList(h1))
	// Expecting [-1, 0, 3, 4, 5]
	printLinkedList(sortList(h2))
	// Expecting []
	printLinkedList(sortList(h3))
}

func printLinkedList(head *ListNode) {
	isFirst := true
	fmt.Printf("[")
	cur := head
	for cur != nil {
		if !isFirst {
			fmt.Printf(", ")
		}
		fmt.Printf("%v", cur.Val)
		cur = cur.Next
		isFirst = false
	}
	println("]")
}

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func sortList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	fakeHead := &ListNode{
		Val:  0,
		Next: head,
	}
	subLen := 1

expansion:
	for {
		prev := fakeHead
		cur := fakeHead.Next
		isMerged := false

		for cur != nil {
			h1 := cur
			cur = skip(cur, subLen)
			h2 := cur
			if h2 == nil && !isMerged {
				break expansion
			}
			cur = skip(cur, subLen)
			mergeResult := mergeSorted(h1, h2, subLen, cur)
			isMerged = true
			prev.Next = mergeResult.Head
			prev = mergeResult.Tail
		}

		subLen <<= 1
	}

	return fakeHead.Next
}

func skip(cur *ListNode, steps int) *ListNode {
	for i := 0; i < steps && cur != nil; i++ {
		cur = cur.Next
	}

	return cur
}

func mergeSorted(h1, h2 *ListNode, length int, next *ListNode) (mergeResult MergeResult) {
	fakeHead := &ListNode{}
	cur := fakeHead
	idx1, idx2 := 0, 0

	for !isConsumed(h1, idx1, length) || !isConsumed(h2, idx2, length) {
		if isConsumed(h1, idx1, length) {
			cur.Next = h2
			h2 = h2.Next
			idx2++
		} else if isConsumed(h2, idx2, length) {
			cur.Next = h1
			h1 = h1.Next
			idx1++
		} else {
			if h1.Val < h2.Val {
				cur.Next = h1
				h1 = h1.Next
				idx1++
			} else {
				cur.Next = h2
				h2 = h2.Next
				idx2++
			}
		}

		cur = cur.Next
	}

	cur.Next = next

	mergeResult.Head = fakeHead.Next
	mergeResult.Tail = cur
	return
}

func isConsumed(cur *ListNode, i, length int) bool {
	return cur == nil || i >= length
}

type MergeResult struct {
	Head *ListNode
	Tail *ListNode
}

File: java/148/Main.java
public class Main {
	public static void main(final String[] args) {
		final var s = new Solution();

		final var h1 = new ListNode(4, new ListNode(2, new ListNode(1, new ListNode(3))));
		final var h2 = new ListNode(-1, new ListNode(5, new ListNode(3, new ListNode(4, new ListNode(0)))));
		final var h3 = (ListNode) null;

		// Expecting [1, 2, 3, 4]
		printLinkedList(s.sortList(h1));
		// Expecting [-1, 0, 3, 4, 5]
		printLinkedList(s.sortList(h2));
		// Expecting []
		printLinkedList(s.sortList(h3));
	}

	private static void printLinkedList(final ListNode head) {
		var isFirst = true;
		System.out.print("[");
		var cur = head;
		while (cur != null) {
			if (!isFirst) {
				System.out.print(", ");
			}
			System.out.print(cur.val);
			cur = cur.next;
			isFirst = false;
		}
		System.out.println("]");
	}
}

/**
 * Definition for singly-linked list.
 */
class ListNode {
	int val;
	ListNode next;

	ListNode() {
	}

	ListNode(final int val) {
		this.val = val;
	}

	ListNode(final int val, final ListNode next) {
		this.val = val;
		this.next = next;
	}
}

class Solution {
	public ListNode sortList(final ListNode head) {
		if (head == null) {
			return null;
		}

		final var fakeHead = new ListNode(0, head);
		var subLen = 1;

		loop: while (true) {
			var prevTail = fakeHead;
			var cur = fakeHead.next;
			var isMerged = false;
			while (cur != null) {
				final var h1 = cur;
				cur = skip(cur, subLen);
				final var h2 = cur;
				if (h2 == null && !isMerged) {
					break loop;
				}
				cur = skip(cur, subLen);
				final var mergeResult = mergeSorted(h1, h2, subLen, cur);
				isMerged = true;
				prevTail.next = mergeResult.getHead();
				prevTail = mergeResult.getTail();
			}

			subLen <<= 1;
		}

		return fakeHead.next;
	}

	private ListNode skip(ListNode cur, final int steps) {
		for (int i = 0; i < steps; i++) {
			if (cur == null) {
				return null;
			}

			cur = cur.next;
		}

		return cur;
	}

	public MergeResult mergeSorted(ListNode h1, ListNode h2, final int length, final ListNode next) {
		final var fakeHead = new ListNode();
		var cur = fakeHead;
		var idx1 = 0;
		var idx2 = 0;

		while (!isConsumed(h1, idx1, length) || !isConsumed(h2, idx2, length)) {
			if (isConsumed(h1, idx1, length)) {
				cur.next = h2;
				h2 = h2.next;
				idx2++;
			} else if (isConsumed(h2, idx2, length)) {
				cur.next = h1;
				h1 = h1.next;
				idx1++;
			} else {
				if (h1.val < h2.val) {
					cur.next = h1;
					h1 = h1.next;
					idx1++;
				} else {
					cur.next = h2;
					h2 = h2.next;
					idx2++;
				}
			}

			cur = cur.next;
		}

		cur.next = next;

		return new MergeResult(fakeHead.next, cur);
	}

	private boolean isConsumed(final ListNode cur, final int i, final int length) {
		return cur == null || i >= length;
	}
}

class MergeResult {
	private final ListNode head;
	private final ListNode tail;

	public MergeResult(final ListNode head, final ListNode tail) {
		this.head = head;
		this.tail = tail;
	}

	public ListNode getHead() {
		return this.head;
	}

	public ListNode getTail() {
		return this.tail;
	}
}
